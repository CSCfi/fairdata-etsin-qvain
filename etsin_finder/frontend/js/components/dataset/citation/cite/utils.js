import moment from 'moment'
import idnToLink from '../../../../utils/idnToLink'

const topOrg = org => {
  if (org.is_part_of) {
    return topOrg(org.is_part_of)
  }
  return org
}

// For surname formatting examples, see
// https://blog.apastyle.org/apastyle/2017/05/whats-in-a-name-two-part-surnames-in-apa-style.html

const startsWithLowerCase = str => str[0] === str[0].toLowerCase()

const suffixParts = ['Jr', 'Jr.', 'Sr', 'Sr.']

const lastnameParts = ['St', 'St.', 'Af', 'Av', 'De', 'Van', 'Von']

const romanNumeralMatch = /^(((IX|IV|V)I{0,3})|I{1,3})$/

const toInitial = name => `${name[0]}.`

export const getNameParts = name => {
  const parts = name.trim().split(' ')
  if (parts.length === 1) {
    return { first: [], last: [parts[0]], suffixes: [] }
  }

  const first = []
  const last = []
  const suffixes = []

  const lastPart = parts[parts.length - 1]
  if (suffixParts.includes(lastPart) || romanNumeralMatch.test(lastPart)) {
    suffixes.push(lastPart)
    parts.splice(parts.length - 1)
  }

  let isFirst = true
  parts.forEach((part, idx) => {
    if (idx === parts.length - 1 || startsWithLowerCase(part) || lastnameParts.includes(part)) {
      isFirst = false
    }
    if (isFirst) {
      first.push(part)
    } else {
      last.push(part)
    }
  })

  return {
    first,
    last,
    suffixes,
  }
}

export const getNameInitials = name => {
  const { first, last, suffixes } = getNameParts(name)

  const components = [last.join(' '), first.map(toInitial).join(' '), suffixes.join(' ')].filter(
    v => v
  )
  return components.join(', ')
}

export const getLastnameFirst = name => {
  const { first, last, suffixes } = getNameParts(name)

  const components = [last.join(' '), first.join(' '), suffixes.join(' ')].filter(v => v)
  return components.join(', ')
}

export const getAuthorsInitial = (dataset, getTranslation, etAlThreshold) => {
  const authors = (dataset.research_dataset.creator || []).map(author => {
    if (author['@type'] === 'Organization') {
      return getTranslation(topOrg(author)?.name)
    }
    return getNameInitials(getTranslation(author.name))
  })
  if (authors.length > etAlThreshold) {
    authors.splice(etAlThreshold - 1, authors.length - etAlThreshold)
    authors[authors.length - 1] = `. . . ${authors[authors.length - 1]}`
  } else if (authors.length > 1) {
    authors[authors.length - 1] = `& ${authors[authors.length - 1]}`
  }
  return authors
}

export const getAuthorsFull = (dataset, getTranslation, etAlThreshold, etAlCount) => {
  const authors = (dataset.research_dataset.creator || []).map((author, index) => {
    if (author['@type'] === 'Organization') {
      return getTranslation(topOrg(author)?.name)
    }
    if (index === 0) {
      return getLastnameFirst(getTranslation(author.name))
    }
    return getTranslation(author.name)
  })
  if (authors.length > etAlThreshold) {
    authors.splice(etAlCount)
    authors[authors.length - 1] = `${authors[authors.length - 1]}, et al.`
  } else if (authors.length > 1) {
    authors[authors.length - 1] = `and ${authors[authors.length - 1]}`
  }
  return authors
}

export const addParens = string => {
  if (string === undefined) {
    return undefined
  }
  return `(${string})`
}

export const addQuotes = string => {
  if (string === undefined) {
    return undefined
  }
  return `”${string}”`
}

export const capitalizeFirst = string => {
  if (string === undefined) {
    return undefined
  }
  return `${string.charAt(0).toUpperCase()}${string.slice(1)}`
}

export const getInitial = string => {
  if (string === undefined) {
    return undefined
  }
  return `${string.charAt(0).toUpperCase()}.`
}

export const getYear = dataset => {
  const issued = dataset.research_dataset.issued
  if (!issued) {
    return undefined
  }
  const year = moment(issued).format('YYYY')
  return year
}

export const getTitle = (dataset, getTranslation) => {
  if (!dataset.research_dataset.title) {
    return undefined
  }
  return getTranslation(dataset.research_dataset.title)
}

export const getVersion = (dataset, getTranslation) => {
  if (!dataset.dataset_version_set) {
    return undefined
  }
  const identifier = dataset.identifier
  const versionIndex = dataset.dataset_version_set.findIndex(v => v.identifier === identifier)
  if (versionIndex < 0) {
    return undefined
  }
  const version = dataset.dataset_version_set.length - versionIndex
  return getTranslation({ en: `Version ${version}`, fi: `versio ${version}` })
}

export const getIdentifier = dataset => {
  const identifier = dataset.research_dataset.preferred_identifier
  if (!identifier) {
    return undefined
  }
  const url = idnToLink(identifier)
  if (url) {
    return url
  }
  if (dataset.draft_of) {
    return dataset.draft_of.preferred_identifier
  }
  if (dataset.state === 'draft') {
    return undefined
  }
  return identifier
}

export const getPublisher = (dataset, getTranslation) => {
  const pub = dataset.research_dataset.publisher
  if (pub) {
    const top = topOrg(pub)
    if (top !== pub) {
      return `${getTranslation(top.name)}, ${getTranslation(pub.name)}`
    }
    return getTranslation(pub.name)
  }
  return undefined
}
