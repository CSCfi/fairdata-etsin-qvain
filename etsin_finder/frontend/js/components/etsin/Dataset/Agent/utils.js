/* eslint-disable camelcase */
export const hasExtraInfo = agent =>
  agent.person || 
  agent.contributor_role || 
  agent.contributor_type || 
  agent.organization.parent || 
  agent.organization.homepage || 
  agent.organization.url

export const flatParentOrgs = (agent) => {
  let org = agent.organization
  const orgs = []
  while (org) {
    orgs.unshift({ ...org })
    org = org.parent
  }
  return orgs
}

export const getOrgKey = org => org.url || Object.values(org.pref_label)[0]

export const getRefDataKey = data => data.identifier || Object.values(data.pref_label)[0]
