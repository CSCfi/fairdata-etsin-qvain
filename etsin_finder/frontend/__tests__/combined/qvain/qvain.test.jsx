import React from 'react'
import { shallow } from 'enzyme'

import '../../../locale/translations'
import { Qvain as QvainBase } from '../../../js/components/qvain/views/main'
import { DATA_CATALOG_IDENTIFIER } from '../../../js/utils/constants'
import { qvainFormSchema } from '../../../js/stores/view/qvain/qvain.submit.schemas'
import { ExternalFilesBase } from '../../../js/components/qvain/fields/files/external/externalFiles'
import DoiSelection, { DoiCheckbox } from '../../../js/components/qvain/fields/files/doiSelection'
import { ButtonGroup } from '../../../js/components/qvain/general/buttons'
import { SlidingContent } from '../../../js/components/qvain/general/card'
import EnvClass from '../../../js/stores/domain/env'
import AccessibilityClass from '../../../js/stores/view/accessibility'
import ElasticQueryClass from '../../../js/stores/view/elasticquery'
import QvainClass, { ExternalResource } from '../../../js/stores/view/qvain'
import LocaleClass from '../../../js/stores/view/locale'
import TablePasState from '../../../js/components/qvain/views/datasets/tablePasState'
import TranslationTab from '../../../js/components/qvain/general/input/translationTab'
import { useStores } from '../../../js/stores/stores'

global.fdweRecordEvent = () => {}

jest.mock('uuid', original => {
  let id = 0
  return {
    ...original,
    v4: () => {
      id += 1
      return id
    },
  }
})

jest.mock('moment', () => {
  const actual = jest.requireActual('moment')
  function momentMock(value) {
    return actual(value || '2020-11-02T12:34Z')
  }
  momentMock.locale = actual.locale
  return momentMock
})

jest.mock('../../../js/stores/stores', () => {
  const mockUseStores = jest.fn()
  return {
    ...jest.requireActual('../../../js/stores/stores'),
    useStores: mockUseStores,
  }
})

const Env = new EnvClass()
const Accessibility = new AccessibilityClass(Env)
const ElasticQuery = new ElasticQueryClass(Env)
const Locale = new LocaleClass(Accessibility, ElasticQuery)
const Qvain = new QvainClass(Env)

const getStores = () => {
  Locale.setLang('en')
  return {
    Env,
    Qvain,
    Locale,
    Matomo: {
      recordEvent: jest.fn(),
    },
  }
}

const emptyMatch = { params: { identifier: null } }

describe('Qvain', () => {
  let stores

  beforeEach(() => {
    stores = getStores()
    stores.Qvain.resetQvainStore()
    useStores.mockReturnValue(stores)
  })

  it('should render correctly', () => {
    const component = shallow(
      <QvainBase Stores={getStores()} history={{}} location={{ pathname: '' }} match={emptyMatch} />
    )

    expect(component).toMatchSnapshot()
  })

  it('should open existing dataset when the identifier in the url changes', () => {
    stores = getStores()

    // Mock react router matches for identifier
    const identifierMatch = { params: { identifier: 'some_identifier' } }
    const anotherMatch = { params: { identifier: 'another_identifier' } }

    // Replace Qvain.getDataset so we can test it was called correctly
    let callCount = 0
    let lastCall
    class FakeQvain extends QvainBase {
      getDataset(identifier) {
        callCount += 1
        lastCall = identifier
      }
    }

    // Create empty dataset, getDataset should not be called
    shallow(
      <FakeQvain Stores={stores} match={emptyMatch} history={{}} location={{ pathname: '/' }} />
    )
    expect(callCount).toBe(0)
    expect(lastCall).toBe(undefined)

    // Dataset already opened for editing (e.g. in the dataset view), getDataset should not be called
    lastCall = undefined
    const datasetOpenedStore = {
      ...stores,
      Qvain: {
        ...stores.Qvain,
        original: { identifier: identifierMatch.params.identifier },
      },
    }
    shallow(
      <FakeQvain
        Stores={datasetOpenedStore}
        match={identifierMatch}
        history={{}}
        location={{ pathname: '/' }}
      />
    )
    expect(callCount).toBe(0)
    expect(lastCall).toBe(undefined)

    // Different dataset already opened, getDataset should be called with the new identifier
    lastCall = undefined
    const anotherDatasetOpenedStore = {
      ...stores,
      Qvain: {
        ...stores.Qvain,
        original: { identifier: anotherMatch.params.identifier },
      },
    }
    shallow(
      <FakeQvain
        Stores={anotherDatasetOpenedStore}
        match={identifierMatch}
        history={{}}
        location={{ pathname: '/' }}
      />
    )
    expect(callCount).toBe(1)
    expect(lastCall).toBe(identifierMatch.params.identifier)

    // Edit existing dataset, getDataset should be called
    lastCall = undefined
    const wrapper = shallow(
      <FakeQvain
        Stores={stores}
        match={identifierMatch}
        history={{}}
        location={{ pathname: '/' }}
      />
    )
    expect(callCount).toBe(2)
    expect(lastCall).toBe(identifierMatch.params.identifier)

    // Switch to another dataset, getDataset should be called
    lastCall = undefined
    wrapper.setProps({ match: anotherMatch })
    expect(callCount).toBe(3)
    expect(lastCall).toBe(anotherMatch.params.identifier)
  })
})

describe('Qvain dataset list PreservationStates', () => {
  let stores

  beforeEach(() => {
    stores = getStores()
    stores.Qvain.resetQvainStore()
    useStores.mockReturnValue(stores)
  })

  it('should render <TablePasState />', () => {
    const component = shallow(<TablePasState preservationState={0} />)
    expect(component).toMatchSnapshot()
  })
})

describe('Qvain translation tabs', () => {
  let stores
  beforeEach(() => {
    stores = getStores()
    useStores.mockReturnValue(stores)
  })

  const getTranslationTabProps = (localeLanguage, activeTabLanguage) => {
    stores.Locale.setLang(localeLanguage)
    const component = shallow(
      <TranslationTab language={activeTabLanguage} setLanguage={() => {}} children="" />
    )
    const langButtons = component.find('[type="button"]').map(btn => btn.props())
    return langButtons.map(({ language, active }) => ({ language, active }))
  }

  it('shows correct TranslationTab', () => {
    expect(getTranslationTabProps('en', 'en')).toEqual([
      { language: 'en', active: true },
      { language: 'fi', active: false },
    ])
    expect(getTranslationTabProps('en', 'fi')).toEqual([
      { language: 'en', active: false },
      { language: 'fi', active: true },
    ])
    expect(getTranslationTabProps('fi', 'fi')).toEqual([
      { language: 'fi', active: true },
      { language: 'en', active: false },
    ])
    expect(getTranslationTabProps('fi', 'en')).toEqual([
      { language: 'fi', active: false },
      { language: 'en', active: true },
    ])
  })
})

describe('Qvain.ExternalFiles', () => {
  let stores

  beforeEach(() => {
    stores = getStores()
    stores.Qvain.resetQvainStore()
    useStores.mockReturnValue(stores)
  })

  it('should render correctly', async () => {
    const externalFiles = shallow(<ExternalFilesBase />)
    expect(externalFiles.find(SlidingContent).length).toBe(1)
    expect(externalFiles.find(ButtonGroup).length).toBe(0)
  })

  // External resources should be listed if there are any
  it('should list all added resources', () => {
    stores.Qvain.saveExternalResource(
      ExternalResource(
        1,
        'External Resource Title',
        'http://en.wikipedia.org',
        'https://en.wikipedia.org/wiki/Portal:Arts'
      )
    )
    const externalFiles = shallow(<ExternalFilesBase />)
    expect(externalFiles.find(ButtonGroup).length).toBe(1)
  })
})

describe('Qvain DOI selection', () => {
  let stores

  beforeEach(() => {
    stores = getStores()
    stores.Qvain.resetQvainStore()
    useStores.mockReturnValue(stores)
  })

  it('should not render DOI selector for ATT catalog', () => {
    stores.Qvain.setDataCatalog(DATA_CATALOG_IDENTIFIER.ATT)
    const component = shallow(<DoiSelection />)
    expect(component.type()).toBe(null)
  })

  it('should not render DOI selector for PAS catalog', () => {
    stores.Qvain.setDataCatalog(DATA_CATALOG_IDENTIFIER.PAS)
    const component = shallow(<DoiSelection />)
    expect(component.type()).toBe(null)
  })

  it('renders DOI selector for new dataset with IDA catalog', () => {
    stores.Qvain.setDataCatalog(DATA_CATALOG_IDENTIFIER.IDA)
    const component = shallow(<DoiSelection />)
    const checkbox = component.find(DoiCheckbox)
    expect(checkbox.prop('checked')).toBe(false)
  })

  it('should not render DOI selector for published dataset', () => {
    const { setDataCatalog, setOriginal } = stores.Qvain
    setDataCatalog(DATA_CATALOG_IDENTIFIER.IDA)
    setOriginal({
      state: 'published',
    })
    const component = shallow(<DoiSelection />)
    expect(component.type()).toBe(null)
  })

  it('renders DOI selector for new draft', () => {
    const { setDataCatalog, setOriginal } = stores.Qvain
    setDataCatalog(DATA_CATALOG_IDENTIFIER.IDA)
    setOriginal({
      state: 'draft',
    })
    const component = shallow(<DoiSelection />)
    const checkbox = component.find(DoiCheckbox)
    expect(checkbox.prop('checked')).toBe(false)
  })

  it('should not render DOI selector for draft of published dataset', () => {
    const { setDataCatalog, setOriginal } = stores.Qvain
    setDataCatalog(DATA_CATALOG_IDENTIFIER.IDA)
    setOriginal({
      state: 'draft',
      draft_of: {
        identifier: 'some_identifier',
      },
    })
    const component = shallow(<DoiSelection />)
    expect(component.type()).toBe(null)
  })

  it('checks the checkbox', () => {
    const { setDataCatalog, setUseDoi } = stores.Qvain
    setDataCatalog(DATA_CATALOG_IDENTIFIER.IDA)
    setUseDoi(true)
    const component = shallow(<DoiSelection />)
    const checkbox = component.find(DoiCheckbox)
    expect(checkbox.prop('checked')).toBe(true)
  })
})

describe('Qvain validation', () => {
  let actors
  let dataset
  beforeEach(() => {
    actors = [
      {
        type: 'organization',
        roles: ['creator'],
        organizations: [
          {
            name: {
              en: 'Test organization',
            },
          },
        ],
      },
    ]
    dataset = {
      title: { en: 'title' },
      description: { en: 'description' },
      keywords: ['keyword'],
      license: [
        {
          name: { en: 'Creative Commons Attribution 4.0 International (CC BY 4.0)' },
          identifier: 'http://uri.suomi.fi/codelist/fairdata/license/code/CC-BY-4.0',
        },
      ],
      dataCatalog: 'urn:nbn:fi:att:data-catalog-ida',
      actors,
      accessType: {
        url: 'http://uri.suomi.fi/codelist/fairdata/access_type/code/open',
      },
    }
  })

  it('should validate dataset', async () => {
    actors[0].roles.push('publisher')
    try {
      expect(await qvainFormSchema.validate(dataset, { abortEarly: false, strict: true }))
    } catch (e) {
      if (e.errors) {
        fail(e.errors)
      } else {
        fail(e)
      }
    }
  })

  it('should fail when required fields for DOI are missing', async () => {
    dataset.useDoi = true // missing publisher role and issuedDate

    try {
      expect(await qvainFormSchema.validate(dataset, { abortEarly: false, strict: true }))
      fail('should have thrown error')
    } catch (e) {
      expect(e.errors.length).toBe(2)
      expect(e.errors).toEqual(
        expect.arrayContaining([
          'qvain.validationMessages.issuedDate.requiredIfUseDoi',
          'qvain.validationMessages.actors.requiredActors.mandatoryActors.publisher',
        ])
      )
    }
  })

  it('should validate dataset with DOI enabled', async () => {
    dataset.useDoi = true
    dataset.issuedDate = '2020-06-01'
    actors[0].roles.push('publisher')

    try {
      expect(await qvainFormSchema.validate(dataset, { abortEarly: false, strict: true }))
    } catch (e) {
      if (e.errors) {
        fail(e.errors)
      } else {
        fail(e)
      }
    }
  })
})
