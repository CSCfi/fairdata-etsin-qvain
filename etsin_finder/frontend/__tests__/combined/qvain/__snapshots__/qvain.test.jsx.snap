// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Qvain dataset list PreservationStates should render <TablePasState /> 1`] = `
<Fragment>
  <Translate
    color="primary"
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": e {
          "baseHash": -703652486,
          "baseStyle": e {
            "baseHash": -1988014698,
            "baseStyle": undefined,
            "componentId": "label__Label-x0jqqw-0",
            "isStatic": false,
            "rules": Array [
              "
  padding: 0.3em 0.6em 0.4em;
  border-radius: 0.2em;
  background: ",
              [Function],
              ";
  color: white;
  display: inline-block;
  margin: ",
              [Function],
              ";
",
            ],
            "staticRulesId": "",
          },
          "componentId": "tablePasState__PasLabel-kh2b5p-0",
          "isStatic": false,
          "rules": Array [
            "
  margin-left: 10px;
  text-transform: uppercase;
",
          ],
          "staticRulesId": "",
        },
        "displayName": "tablePasState__PasLabel",
        "foldedComponentIds": Array [
          "label__Label-x0jqqw-0",
        ],
        "render": [Function],
        "shouldForwardProp": undefined,
        "styledComponentId": "tablePasState__PasLabel-kh2b5p-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    content="qvain.files.dataCatalog.pas"
  />
  <Translate
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": e {
          "baseHash": -1476272384,
          "baseStyle": undefined,
          "componentId": "tablePasState__PasText-kh2b5p-1",
          "isStatic": false,
          "rules": Array [
            "
  margin-left: 10px;
  font-size: 0.8em;
  :before {
    content: '(';
  }
  :after {
    content: ')';
  }
",
          ],
          "staticRulesId": "",
        },
        "displayName": "tablePasState__PasText",
        "foldedComponentIds": Array [],
        "render": [Function],
        "shouldForwardProp": undefined,
        "styledComponentId": "tablePasState__PasText-kh2b5p-1",
        "target": "span",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    content="qvain.pasState.0"
  />
</Fragment>
`;

exports[`Qvain should render correctly 1`] = `
<card__QvainContainer>
  <Header
    datasetError={false}
    datasetLoading={false}
  />
  <StickyHeader
    clearSubmitResponse={[Function]}
    datasetError={false}
    datasetLoading={false}
    handleSubmitError={[Function]}
    handleSubmitResponse={[Function]}
    hideSubmitButtons={false}
    response={null}
    submitButtonsRef={
      Object {
        "current": null,
      }
    }
    submitted={false}
  />
  <ErrorBoundary
    ContainerComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [
          Object {
            "className": "error",
          },
        ],
        "componentStyle": e {
          "baseHash": -342646166,
          "baseStyle": undefined,
          "componentId": "errorBoundary__ErrorContainer-kebdqo-0",
          "isStatic": false,
          "rules": Array [
            "
  && {
    padding-left: 2rem;
    padding-right: 2rem;
  }
",
          ],
          "staticRulesId": "",
        },
        "displayName": "errorBoundary__ErrorContainer",
        "foldedComponentIds": Array [],
        "render": [Function],
        "shouldForwardProp": undefined,
        "styledComponentId": "errorBoundary__ErrorContainer-kebdqo-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    callback={[Function]}
    title={
      <Translate
        component="h2"
        content="qvain.error.render"
      />
    }
  >
    <Dataset
      datasetError={false}
      datasetErrorDetails={null}
      datasetErrorTitle={null}
      handleRetry={[Function]}
      handleSubmitError={[Function]}
      handleSubmitResponse={[Function]}
      haveDataset={true}
      setFocusOnSubmitButton={[Function]}
    />
  </ErrorBoundary>
  <LooseActorsDialog />
  <LooseProvenanceDialog />
  <Translate
    attributes={
      Object {
        "message": "qvain.unsavedChanges",
      }
    }
    component={[Function]}
    when={false}
  />
</card__QvainContainer>
`;

exports[`Qvain.Description should render <DescriptionField /> 1`] = `
<div
  className="container"
>
  <title__SectionTitle>
    <Translate
      component="span"
      content="qvain.description.title"
    />
    <Tooltip
      align="Right"
      close={[Function]}
      isOpen={false}
      text={<DescriptionInfo />}
    >
      <HelpIcon
        aria-label="Description info"
        onClick={[Function]}
      />
    </Tooltip>
  </title__SectionTitle>
  <withFieldErrorBoundary(DescriptionField) />
  <withFieldErrorBoundary(IssuedDateField) />
  <withFieldErrorBoundary(OtherIdentifierField) />
  <withFieldErrorBoundary(FieldOfScienceField) />
  <withFieldErrorBoundary(LanguageField) />
  <withFieldErrorBoundary(KeywordsField) />
  <withFieldErrorBoundary(SubjectHeadingsField) />
</div>
`;

exports[`Qvain.Description should render <FieldOfScienceField /> 1`] = `
<Card
  bottomContent={false}
>
  <form__LabelLarge
    htmlFor="fieldOfScience-select"
  >
    <Translate
      component="span"
      content="qvain.description.fieldOfScience.title"
    />
  </form__LabelLarge>
  <Translate
    component="p"
    content="qvain.description.fieldOfScience.help"
  />
  <Translate
    component={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      }
    }
    getRefGroups={[Function]}
    getter={Array []}
    isClearable={false}
    isMulti={true}
    metaxIdentifier="field_of_science"
    model={[Function]}
    modifyGroupLabel={[Function]}
    modifyOptionLabel={[Function]}
    name="fieldOfScience"
    placeholder="qvain.description.fieldOfScience.placeholder"
    setter={[Function]}
    sortFunc={[Function]}
  />
</Card>
`;

exports[`Qvain.Description should render <KeywordsField /> 1`] = `
<Card
  bottomContent={false}
>
  <form__LabelLarge
    htmlFor="keywords-input"
  >
    <TooltipHover
      flexGrow={0}
      position="right"
      showOnClick={false}
      showOnHover={true}
      title="Required to save and publish"
    >
      <Translate
        component="span"
        content="qvain.description.keywords.title"
      />
       *
    </TooltipHover>
  </form__LabelLarge>
  <Translate
    component="p"
    content="qvain.description.keywords.help"
  />
  <StringArray
    addItemStr={[Function]}
    addWithComma={true}
    id="keywords-input"
    itemSchema={null}
    itemStr=""
    readonly={false}
    required={true}
    schema={
      ArraySchema {
        "_blacklist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_blacklistError": undefined,
        "_mutate": undefined,
        "_typeError": [Function],
        "_whitelist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_whitelistError": undefined,
        "conditions": Array [],
        "deps": Array [],
        "exclusiveTests": Object {
          "required": true,
        },
        "innerType": StringSchema {
          "_blacklist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_blacklistError": undefined,
          "_typeError": [Function],
          "_whitelist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_whitelistError": undefined,
          "conditions": Array [],
          "deps": Array [],
          "exclusiveTests": Object {
            "max": true,
          },
          "spec": Object {
            "abortEarly": true,
            "nullable": false,
            "presence": "optional",
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "spec": Object {
          "abortEarly": true,
          "nullable": false,
          "presence": "required",
          "recursive": true,
          "strict": false,
          "strip": false,
        },
        "tests": Array [
          [Function],
        ],
        "transforms": Array [
          [Function],
        ],
        "type": "array",
      }
    }
    set={[Function]}
    setItemStr={[Function]}
    setValidationError={[Function]}
    translationsRoot="qvain.description.keywords"
    validationError=""
    value={Array []}
  />
</Card>
`;

exports[`Qvain.Description should render <LanguageField /> 1`] = `
<Card
  bottomContent={false}
>
  <form__LabelLarge
    htmlFor="dataset-language-select"
  >
    <Translate
      component="span"
      content="qvain.description.datasetLanguage.title"
    />
  </form__LabelLarge>
  <Translate
    component="p"
    content="qvain.description.datasetLanguage.help"
  />
  <SearchSelect
    getter={Array []}
    id="datasetLanguage"
    inModal={false}
    isClearable={false}
    isMulti={true}
    metaxIdentifier="language"
    model={[Function]}
    name="dataset-language"
    noOptionsMessage={[Function]}
    placeholder="qvain.description.datasetLanguage.placeholder"
    search={true}
    setter={[Function]}
  />
</Card>
`;

exports[`Qvain.Description should render <OtherIdentifierField /> 1`] = `
<Card
  bottomContent={true}
>
  <form__LabelLarge
    htmlFor="other-identifiers-input"
  >
    <Translate
      component="span"
      content="qvain.description.otherIdentifiers.title"
    />
  </form__LabelLarge>
  <Translate
    component="p"
    content="qvain.description.otherIdentifiers.instructions"
  />
  <StringArray
    addItemStr={[Function]}
    addWithComma={true}
    id="other-identifiers-input"
    itemSchema={
      StringSchema {
        "_blacklist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_blacklistError": undefined,
        "_typeError": [Function],
        "_whitelist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_whitelistError": undefined,
        "conditions": Array [],
        "deps": Array [],
        "exclusiveTests": Object {
          "max": true,
          "min": true,
          "url": false,
        },
        "spec": Object {
          "abortEarly": true,
          "nullable": false,
          "presence": "optional",
          "recursive": true,
          "strict": false,
          "strip": false,
        },
        "tests": Array [
          [Function],
          [Function],
          [Function],
        ],
        "transforms": Array [
          [Function],
        ],
        "type": "string",
      }
    }
    itemStr=""
    readonly={false}
    required={false}
    schema={
      ArraySchema {
        "_blacklist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_blacklistError": undefined,
        "_typeError": [Function],
        "_whitelist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_whitelistError": undefined,
        "conditions": Array [],
        "deps": Array [],
        "exclusiveTests": Object {},
        "innerType": StringSchema {
          "_blacklist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_blacklistError": undefined,
          "_typeError": [Function],
          "_whitelist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_whitelistError": undefined,
          "conditions": Array [],
          "deps": Array [],
          "exclusiveTests": Object {
            "max": true,
            "min": true,
            "url": false,
          },
          "spec": Object {
            "abortEarly": true,
            "nullable": false,
            "presence": "optional",
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "spec": Object {
          "abortEarly": true,
          "nullable": true,
          "presence": "optional",
          "recursive": true,
          "strict": false,
          "strip": false,
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
        "type": "array",
      }
    }
    set={[Function]}
    setItemStr={[Function]}
    setValidationError={[Function]}
    translationsRoot="qvain.description.otherIdentifiers"
    validationError=""
    value={Array []}
  />
</Card>
`;

exports[`Qvain.Description should render <SubjectHeadingsField /> 1`] = `
<Card
  bottomContent={false}
>
  <form__LabelLarge
    htmlFor="subjectHeading-select"
  >
    <Translate
      component="span"
      content="qvain.description.subjectHeadings.title"
    />
  </form__LabelLarge>
  <Translate
    component="p"
    content="qvain.description.subjectHeadings.help"
  />
  <SearchSelect
    getter={Array []}
    inModal={false}
    isClearable={false}
    isMulti={true}
    metaxIdentifier="keyword"
    model={[Function]}
    name="subjectHeading"
    placeholder="qvain.description.subjectHeadings.placeholder"
    setter={[Function]}
  />
</Card>
`;

exports[`Qvain.RightsAndLicenses should render <AccessType /> 1`] = `
<Card
  bottomContent={false}
>
  <form__LabelLarge
    htmlFor="accessTypeSelect"
  >
    <Translate
      component="span"
      content="qvain.rightsAndLicenses.accessType.title"
    />
  </form__LabelLarge>
  <Translate
    attributes={
      Object {
        "placeholder": "qvain.rightsAndLicenses.accessType.placeholder",
      }
    }
    clearable={true}
    component={[Function]}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    inputId="accessTypeSelect"
    isDisabled={false}
    name="accessType"
    onBlur={[Function]}
    onChange={[Function]}
    options={Array []}
    value={
      Object {
        "name": undefined,
        "url": "http://uri.suomi.fi/codelist/fairdata/access_type/code/open",
      }
    }
  />
  <ValidationError />
</Card>
`;

exports[`Qvain.RightsAndLicenses should render <Licenses /> 1`] = `
<Card
  bottomContent={false}
>
  <form__LabelLarge
    htmlFor="licenseSelect"
  >
    <Translate
      component="span"
      content="qvain.rightsAndLicenses.license.title"
    />
  </form__LabelLarge>
  <Translate
    component="p"
    content="qvain.rightsAndLicenses.license.infoText"
  />
  <Translate
    attributes={
      Object {
        "placeholder": "qvain.rightsAndLicenses.license.placeholder",
      }
    }
    component={[Function]}
    createOptionPosition="first"
    getNewOptionData={[Function]}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    inputId="licenseSelect"
    isClearable={false}
    isDisabled={false}
    isMulti={true}
    name="license"
    onChange={[Function]}
    options={Array []}
    styles={
      Object {
        "multiValue": [Function],
        "multiValueLabel": [Function],
      }
    }
    value={
      Array [
        Object {
          "identifier": "http://uri.suomi.fi/codelist/fairdata/license/code/CC-BY-4.0",
          "name": undefined,
        },
      ]
    }
  />
  <ValidationErrors
    errors={Array []}
  />
</Card>
`;

exports[`Qvain.RightsAndLicenses should render <RightsAndLicenses /> 1`] = `
<div
  className="container"
>
  <title__SectionTitle>
    <Translate
      component="span"
      content="qvain.rightsAndLicenses.title"
    />
    <Tooltip
      align="Right"
      close={[Function]}
      isOpen={false}
      text={<LicensesInfo />}
    >
      <HelpIcon
        aria-label="Rights And Licenses info"
        onClick={[Function]}
      />
    </Tooltip>
  </title__SectionTitle>
  <withFieldErrorBoundary(WithTheme(Component)) />
  <withFieldErrorBoundary(Component) />
</div>
`;
